rpc/src/rpc.rs

```diff
@@ -605,30 +647,87 @@ impl JsonRpcRequestProcessor {
+                        || (!partitioned_epoch_reward_enabled && reward_type == RewardType::Staking)
+                },
+            )
+        };
 
-        let reward_hash: HashMap<String, Reward> = first_confirmed_block
-            .rewards
-            .unwrap_or_default()
-            .into_iter()
-            .filter_map(|reward| match reward.reward_type? {
-                RewardType::Staking | RewardType::Voting => addresses
-                    .contains(&reward.pubkey)
-                    .then(|| (reward.clone().pubkey, reward)),
-                _ => None,
-            })
-            .collect();
+        if partitioned_epoch_reward_enabled {
+            let num_partitions = epoch_boundary_block.num_reward_partitions.expect(
+                "epoch-boundary block should have num_reward_partitions after partitioned epoch \
+                 rewards enabled",
+            );
+
+            let num_partitions = usize::try_from(num_partitions)
+                .expect("num_partitions should never exceed usize::MAX");
+            let hasher = EpochRewardsHasher::new(
+                num_partitions,
+                &Hash::from_str(&epoch_boundary_block.previous_blockhash)
+                    .expect("UiConfirmedBlock::previous_blockhash should be properly formed"),
+            );
+            let mut partition_index_addresses: HashMap<usize, Vec<String>> = HashMap::new();
+            for address in addresses.iter() {
+                let address_string = address.to_string();
+                // Skip this address if (Voting) rewards were already found in
+                // the first block of the epoch
+                if !reward_map.contains_key(&address_string) {
+                    let partition_index = hasher.clone().hash_address_to_partition(address);
+                    partition_index_addresses
+                        .entry(partition_index)
+                        .and_modify(|list| list.push(address_string.clone()))
+                        .or_insert(vec![address_string]);
+                }
+            }
+
+            let block_list = self
+                .get_blocks_with_limit(
+                    first_confirmed_block_in_epoch + 1,
+                    num_partitions,
+                    Some(context_config),
+                )
+                .await?;
+
+            for (partition_index, addresses) in partition_index_addresses.iter() {
+                let slot = *block_list
+                    .get(*partition_index)
+                    .ok_or_else(Error::internal_error)?;
+
+                let Ok(Some(block)) = self
+                    .get_block(
+                        slot,
+                        Some(RpcBlockConfig::rewards_with_commitment(config.commitment).into()),
+                    )
+                    .await
+                else {
+                    return Err(RpcCustomError::BlockNotAvailable { slot }.into());
+                };
```

Comment:
Are you thinking a custom error? Or...?
I'm sure you figured this out since you put the comment where you did, but if the query lands when any stake-account address hasn't been rewarded yet, RPC will return BlockNotAvailable error for the whole query. I suppose we could skip this error and return `None` for the specific addresses instead, but then there's no indication why those rewards aren't there, and when they will be available.

Reply:
It may not be a huge deal either way. An intelligent client could construct the EpochRewardsHasher itself and use `min_context_slot` accordingly.

Reply:
intelligent client sounds pretty optimistic 😅 

wdyt about collecting the "pending rewards" addresses to a separate field in the reply? something like
```
pending_rewards: [
{
  slot: SLOT,
  addresses: [
    ADDR1,
    ADDR2,
    ...
  ]
},
...
]
```

Reply:
That's a nice idea, but unfortunately a much bigger change, since currently the result field is a list of rewards that matches the order of the input address-list.
We could, however, add something like a `pending` field to `RpcInflationReward`, and zero out the balance fields: https://github.com/anza-xyz/agave/blob/3ee204b5cba93211fa0e07dc9238372eeb7933db/rpc-client-api/src/response.rs#L511-L517
Unfortunately, we don't know the effective_slot for sure yet, so I guess that would have to be zero also.

Reply:
ugh... yeah good point(s). let's punt on the details and add a more specific error code? `RpcCustomError::RewardsDistributionInProgress` or similar?

Reply:
Pushed. See what you think @t-nelson

Reply:
Here's an example output:
```
$ curl  -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0", "id":1, "method":"getInflationReward", "params":[["2VM7iZ76bCneebPdDqLCqSmHeRhhdJpXCMMeBCSG4Asz","z8MHeH6EaBjRmXM3Y5qJ7LtLGzdvv3UaUQzXxqyQLdS"],{"epoch":4,"commitment":"confirmed"}]}' 127.0.0.1:8899 | jq
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32017,
    "message": "Epoch rewards period still active at slot 321",
    "data": {
      "currentBlockHeight": 321,
      "rewardsCompleteBlockHeight": 327
    }
  },
  "id": 1
}
```
The error is a little misleading, because if all the requested rewards are actually available at some point before `rewardsCompleteBlockHeight`, the method will happily return them.